lte/gateway/c/session_manager/LocalEnforcer.cpp

```diff
@@ -1016,13 +1016,44 @@ void LocalEnforcer::init_session(
     const std::string& session_id, const SessionConfig& cfg,
     const CreateSessionResponse& response) {
   const auto time_since_epoch = magma::get_time_in_sec_since_epoch();
-  unsigned shard_id           = shard_tracker_->add_ue(imsi);
   auto session                = std::make_unique<SessionState>(
       imsi, session_id, cfg, *rule_store_, response.tgpp_ctx(),
-      time_since_epoch, response, shard_id);
+      time_since_epoch, response);
   session_map[imsi].push_back(std::move(session));
 }
 
+void LocalEnforcer::add_ue_to_shard(
+    const std::string imsi, SessionMap& session_map, SessionState& session) {
+  // Check if UE has other sessions and reuse shard_id if so
+  auto sm_it     = session_map.find(imsi);
+  auto& sessions = sm_it->second;
+  for (auto it = sessions.begin(); it != sessions.end(); ++it) {
+    if ((*it)->get_session_id() != session.get_session_id()) {
+      session.set_shard_id((*it)->get_shard_id());
+      break;
+    }
+  }
+  // if the UE doesn't have any sessions, add to shard tracker and set id
+  if (sm_it == session_map.end()) {
+    int shard_id = shard_tracker_->add_ue(imsi);
+    session.set_shard_id(shard_id);
+  }
+}
+
+bool LocalEnforcer::remove_ue_from_shard(
+    const std::string imsi, SessionMap& session_map, SessionState& session) {
+  auto session_map_it = session_map.find(imsi);
```

Comment:
Let's return early if there are no sessions for the IMSI

Reply:
So it shouldn't ever get called if there are no sessions for the IMSI(since it's dependent on the session existing in the map during invocation in update tunnel ids).